File: LeetCode Solutions/May 2020 LeetCoding Challenges/Search Suggestions System.java
class Solution {
    public List<List<String>> suggestedProducts(String[] products, String searchWord) {
        Arrays.sort(products);
        List<List<String>> List = new ArrayList<>();
        int left = 0, right = products.length - 1;
        for(int i=0; i<searchWord.length(); i++){
            List<String> result = new ArrayList<>();
            char c = searchWord.charAt(i);
            while(left <= right && (products[left].length() == i || products[left].charAt(i) < c)){
                left++;
            }
            while(left <= right && (products[right].length() == i || products[right].charAt(i) > c)){
                right--;
            }
            for(int j=0; j<3 && left+j<=right; j++){
                result.add(products[left+j]);
            }
            List.add(result);
        }
        return List;
    }
}

File: LeetCode Solutions/LeetCoding Challenge/May 2020 LeetCoding Challenge/Search Suggestions System.py
class Solution:
    def suggestedProducts(self, products: List[str], searchWord: str) -> List[List[str]]:
        products.sort()
        List, left, right = [], 0, len(products)-1
        for i in range(len(searchWord)):
            count, result = searchWord[i], []
            while(left <= right and (len(products[left]) == i or products[left][i] < count)):
                  left += 1
            while(left <= right and (len(products[right]) == i or products[right][i] > count)):
                right -= 1
            for j in range(3):
                if(left + j > right):
                    break
                else:
                    result.append(products[left+j])
            List.append(result)
        return List